apps/aecore/src/aec_chain_state.erl

```diff
@@ -898,18 +930,33 @@ db_children(#node{} = Node) ->
             ++ aec_db:find_headers_at_height(Height),
         aec_headers:prev_hash(Header) =:= Hash].
 
-db_node_has_sibling_blocks(Node) ->
+db_sibling_blocks(#node{type = key} = Node) ->
     Height   = node_height(Node),
     PrevHash = prev_hash(Node),
-    %% NOTE: Micro blocks have the same height.
-    %% For key blocks siblings at Height and Height - 1
-    %% and for micro blocks siblings at Height and Height + 1
-    OtherHeight =
-        case node_type(Node) of
-            micro -> Height + 1;
-            key   -> Height - 1
-        end,
-    length([x || Header <- aec_db:find_headers_at_height(Height)
-                           ++ aec_db:find_headers_at_height(OtherHeight),
-                 aec_headers:prev_hash(Header) =:= PrevHash]) > 1.
+    #{ key_sibling => match_prev_at_height(Height, PrevHash),
+       micro_sibling => match_prev_at_height(Height-1, PrevHash)};
+db_sibling_blocks(#node{type = micro} = Node) ->
+    Height   = node_height(Node),
+    PrevHash = prev_hash(Node),
+    #{ key_sibling => match_prev_at_height(Height+1, PrevHash),
+       micro_sibling => match_prev_at_height(Height, PrevHash)}.
 
+match_prev_at_height(Height, PrevHash) ->
+    [Header || Header <- aec_db:find_headers_at_height(Height),
+               aec_headers:prev_hash(Header) =:= PrevHash].
+
+db_mark_malicious_leader(Node) ->
+    Height = node_height(Node),
+    {ok, MaliciousKeyBlock} = aec_chain:get_key_block_by_height(Height - 1),
+    MaliciousHeader         = aec_blocks:to_header(MaliciousKeyBlock),
+    {ok, MaliciousHash}     = aec_headers:hash_header(MaliciousHeader),
+
+    {value, Trees, Difficulty, ForkId, Fees, KeyHash, _Fraud}
```

Comment:
We may want to consider separate structure for fraud key blocks, without extension of `#fork_info{}`.